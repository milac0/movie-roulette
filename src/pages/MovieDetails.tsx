import React, { useEffect, useState, useContext } from "react";
import { useParams } from "react-router-dom";
import { css } from "emotion";
import axios from "axios";
import {
  getGenres,
  getRuntime,
  getProductionCompanies,
  getYear
} from "../helpers";

import { UserContext } from "./../context/UserContext";

import CustomRating from "../components/CustomRating";

const details = css`
  background: #eeeeee;
  margin: 0 auto;
  padding: 1em;
  width: 660px;
  & .title-block {
    background: #333333;
    color: #fff;
    margin-bottom: 1em;
    padding: 1em;
    & h1 {
      font-size: 2rem;
      font-weight: 400;
      margin: 0;
    }
    & h2 {
      font-size: 1rem;
      font-weight: 400;
      margin: 0;
      margin-bottom: 0.75em;
    }
  }
  .rating {
    display: flex;
    justify-content: space-between;
    & p {
      font-size: 2rem;
      margin: 0;
    }
    & .rating-outof {
      font-size: 1rem;
    }
  }
  & p {
    font-size: 1rem;
  }
`;

const image = css`
  display: flex;
  justify-content: space-between;
  & .image-poster {
    margin-right: 1em;
    height: 281px;
  }
`;

interface Props {}

const MovieDetails: React.FC<Props> = () => {
  const { movieid } = useParams();
  const [movie, setMovie] = useState();

  useEffect(() => {
    (async () => {
      const response = await axios.get(
        `movie/${movieid}?api_key=${process.env.API_KEY}&language=en-US`
      );
      setMovie(response.data);
    })();
  }, []);

  return (
    <div>
      {movie === undefined ? (
        <h1>Loading skeleton</h1>
      ) : (
        <div className={details}>
          <div className="title-block">
            <h1>
              {movie.title} ({getYear(movie.release_date)}){" "}
            </h1>
            <h2>
              {getRuntime(movie.runtime)} &nbsp; | &nbsp;{" "}
              {getGenres(movie.genres)} &nbsp; | &nbsp;{" "}
              {movie.original_language}
            </h2>
            <div className={image}>
              <img
                className="image-poster"
                src={`https://image.tmdb.org/t/p/w500${movie.poster_path}`}
              />
              <img
                className="image-backdrop"
                src={`https://image.tmdb.org/t/p/w500${movie.backdrop_path}`}
              />
            </div>
          </div>
          <div className="rating">
            <div>
              <p>
                {movie.vote_average}
                <span className="rating-outof">/10</span>
              </p>
              <span>{movie.popularity}</span>
            </div>
            <CustomRating movieid={movieid} />
          </div>
          <p>{movie.overview}</p>
          <p>
            Production companies:{" "}
            {getProductionCompanies(movie.production_companies)}
          </p>
        </div>
      )}
    </div>
  );
};

export default MovieDetails;
