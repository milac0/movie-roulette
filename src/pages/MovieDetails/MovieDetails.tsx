import React, { useEffect, useState } from "react";
import { useParams } from "react-router-dom";
import axios from "axios";
import {
  getGenres,
  getRuntime,
  getProductionCompanies
} from "./movieDetails.helpers";
import { getYear } from "../../helpers/index";
import CustomRating from "../../components/CustomRating/CustomRating";
import styles from "./movieDetails.scss";
import Movie from "../../interface/Movie";
import ScrollToTopOnMount from "../../components/ScrollToTopOnMount";

interface Props {}

const MovieDetails: React.FC<Props> = () => {
  const { movieid } = useParams();
  const movieURL = `/movie/${movieid}?api_key=${process.env.API_KEY}&language=en-US`;
  const trailerURL = `movie/${movieid}/videos?api_key=${process.env.API_KEY}&language=en-US`;
  const [movie, setMovie] = useState<Movie | null>(null);
  const [trailer, setTrailer] = useState(null);

  useEffect(() => {
    (async () => {
      const {
        data: { results }
      } = await axios.get(trailerURL);
      if (results.length > 0 && results[0].key !== null) {
        setTrailer(results[0].key);
      }
      const { data: movie } = await axios.get(movieURL);
      setMovie(movie);
    })();
  }, []);

  return (
    <div>
      {!movie ? (
        <div className={styles.skeleton}></div>
      ) : (
        <div className={styles.details}>
          <ScrollToTopOnMount />
          <div className={styles.titleBlock}>
            <h1>
              {movie.title} ({getYear(movie.release_date)}){" "}
            </h1>
            <h2>
              {getRuntime(movie.runtime)} &nbsp; | &nbsp;{" "}
              {getGenres(movie.genres)} &nbsp; | &nbsp;{" "}
              {movie.original_language}
            </h2>
            <div className={styles.image}>
              <img
                className={styles.imagePoster}
                src={
                  movie.poster_path
                    ? `https://image.tmdb.org/t/p/w500${movie.poster_path}`
                    : "http://via.placeholder.com/187x281"
                }
              />
              {trailer ? (
                <div className={styles.responsiveContainer}>
                  <iframe
                    src={`https://www.youtube.com/embed/${trailer}`}
                    allowFullScreen
                  />
                </div>
              ) : (
                <img
                  className={styles.imageBackdrop}
                  src={
                    movie.backdrop_path
                      ? `https://image.tmdb.org/t/p/w500${movie.backdrop_path}`
                      : "http://via.placeholder.com/425x281"
                  }
                />
              )}
            </div>
          </div>
          <div className={styles.rating}>
            <div>
              <p>
                {movie.vote_average}
                <span>/10</span>
              </p>
              <span className={styles.popularity}>{movie.popularity}</span>
            </div>
            <CustomRating movieid={parseInt(movieid)} />
          </div>
          <p>{movie.overview}</p>
          <p>
            Production companies:{" "}
            {getProductionCompanies(movie.production_companies)}
          </p>
        </div>
      )}
    </div>
  );
};

export default MovieDetails;
