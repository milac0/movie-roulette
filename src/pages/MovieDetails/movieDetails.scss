@import "../../theme/variables";

.details {
  background: $backgroundGrey;
  margin: 0 auto;
  margin-bottom: 2em;
  padding: 1em;
  text-align: center;

  .titleBlock {
    background: #333333;
    color: #fff;
    margin-bottom: 1em;
    padding: 1em;

    h1 {
      font-size: 1.5rem;
      font-weight: 400;
      margin: 0;
    }

    h2 {
      font-size: 1rem;
      font-weight: 400;
      margin: 0;
      margin-bottom: 0.75em;
    }
  }

  .rating {
    display: flex;
    flex-direction: column-reverse;

    p {
      font-size: 2rem;
      margin: 0;
      text-align: center;

      span {
        font-size: 1rem;
      }
    }

    .popularity {
      font-size: 1rem;
    }
  }

  p {
    font-size: 1rem;
    text-align: left;
  }
}

.image {
  .imagePoster {
    margin-bottom: 0.5em;
    width: 100%;
  }

  .imageBackdrop {
    width: 90%;
  }

  .responsiveContainer {
    position: relative;
    padding-bottom: 56.25%;
    padding-top: 25px;
    height: 0;

    iframe {
      border: 0;
      position: absolute;
      top: 0;
      left: 0;
      width: 100%;
      height: 100%;
    }
  }
}

.skeleton {
  height: 600px;
  width: 700px;
  margin: 0 auto;
}

@media (min-width: 414px) {
  .details {
    margin-bottom: 5em;
    text-align: left;
    width: 660px;

    .titleBlock {
      text-align: start;

      h1 {
        font-size: 2rem;
      }
    }

    .rating {
      flex-direction: row;
      justify-content: space-between;
    }

    .image {
      display: flex;
      justify-content: space-evenly;

      .imagePoster {
        height: 281px;
        margin-right: 1em;
        width: 220px;
      }

      .imageBackdrop {
        height: 281px;
        width: 425px;
      }

      .responsiveContainer {
        padding: 0;
        position: static;

        iframe {
          position: static;
          border: 0;
          height: 281px;
          width: 385px;
        }
      }
    }
  }
}
