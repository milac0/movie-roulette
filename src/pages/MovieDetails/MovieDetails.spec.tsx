import React from 'react';
import renderer from 'react-test-renderer';

import MovieDetails from './MovieDetails';

jest.mock('react-router', () => ({
    useParams: jest.fn().mockReturnValue({ movieid: '419704' }),
}));

describe('MovieDetails renders correctly', () => {
    test('snapshot renders', () => {
        const tree = renderer.create(<MovieDetails />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});

