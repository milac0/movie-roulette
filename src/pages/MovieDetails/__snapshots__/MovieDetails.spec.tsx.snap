// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetails renders correctly snapshot renders 1`] = `
<div>
  <div
    className="skeleton"
  />
</div>
`;
